---
title: "Lymphocyte subset analysis without TCR"
author: "Vivian Lau"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r library and object setup}
library(knitr)
library(Seurat)
library(tidyverse)
library(ggplot2)
library(cowplot)
library(patchwork)
library(fgsea)
library(msigdbr)
library(ggrepel)
library(biomaRt)
```


```{r}
Lymphocytes <- readRDS("~/Documents/Bioinformatics/sc_analysis_2/Rds/Lymphocytes.rds")

# Mitochondrial genes
Lymphocytes[["percent.mt"]] <- PercentageFeatureSet(Lymphocytes, pattern = "^Mt-")
# Ribosomal proteins
Lymphocytes[["percent.ribo"]] <- PercentageFeatureSet(Lymphocytes, pattern = "^Rp[sl]")

# Removing TCR genes
Lymphocytes <- Lymphocytes[!grepl("^Tr[abdg][vjc]", rownames(Lymphocytes)), ]

# Normalization, scaling and initial clustering
Lymphocytes <- NormalizeData(Lymphocytes, normalization.method = "LogNormalize", scale.factor = 10000)
Lymphocytes <- FindVariableFeatures(Lymphocytes, selection.method = "vst", nfeatures = 2000)
Lymphocytes <- ScaleData(Lymphocytes, features = rownames(Lymphocytes), vars.to.regress = c("percent.mt","percent.ribo"))
Lymphocytes <- RunPCA(Lymphocytes, features = VariableFeatures(object = Lymphocytes))
Lymphocytes <- RunUMAP(Lymphocytes, dims= 1:10)
ElbowPlot(object = Lymphocytes)
Lymphocytes <- FindNeighbors(Lymphocytes, dims = 1:10, nn.method = "annoy")
Lymphocytes <- FindClusters(Lymphocytes, resolution = c(0.4))

# # Single cell transform
# Lymphocytes <- SCTransform(Lymphocytes, verbose = FALSE, ncells = NULL)
# Lymphocytes <- RunPCA(Lymphocytes, verbose = FALSE)
# ElbowPlot(object = Lymphocytes)
# Lymphocytes <- RunUMAP(Lymphocytes, dims = 1:20, verbose = FALSE)
# Lymphocytes <- FindNeighbors(Lymphocytes, dims = 1:20, verbose = FALSE, nn.method = "rann")
# Lymphocytes <- FindClusters(Lymphocytes, resolution = c(seq(from = 0.1, to = 2.0, by = 0.1)))
# clustree(x = Lymphocytes, prefix = "SCT_snn_res.")
# Lymphocytes <- FindClusters(Lymphocytes, verbose = FALSE, resolution = c(0.4))

DimPlot(Lymphocytes, reduction = "umap", group.by = "seurat_clusters", split.by = "model")

# Add TCR-genes back after clustering
Lymphocytes_wTCR <- readRDS("~/Library/CloudStorage/OneDrive-Nexus365/Oxford/IITM/DPhil Data/Single Cell/02_VLCD45_GEX_HTO/analysis/Rds/Lymphocytes.rds")

a <- subset(Lymphocytes_wTCR, cells = WhichCells(object = Lymphocytes))
Lymphocytes[["new.counts"]] <- CreateAssayObject(counts = GetAssayData(object = a, slot = "count", assay = "RNA"))
Lymphocytes <- NormalizeData(Lymphocytes, normalization.method = "LogNormalize", scale.factor = 10000, assay = "new.counts")
Lymphocytes <- ScaleData(Lymphocytes, features = rownames(Lymphocytes), assay = "new.counts")

Lymphocytes <- subset(Lymphocytes, ident = 0:6)
DimPlot(Lymphocytes, reduction = "umap", group.by = "seurat_clusters", split.by = "model")

## Top-10 upregulated DEGs
DefaultAssay(Lymphocytes) <- "RNA"
Lymphocytes <- NormalizeData(object = Lymphocytes, scale.factor = 10000)
Lymphocytes <- ScaleData(object = Lymphocytes, features = rownames(Lymphocytes))

Lymphocytes.markers <- FindAllMarkers(object = Lymphocytes, logfc.threshold = 0.25, Lymphocytes.use = "negbinom", only.pos = T, return.thresh = 0.05)

Lymphocytes.markers.filtered <- Lymphocytes.markers %>% filter(p_val_adj > 0 & p_val_adj < 0.05)

Lymphocytes.markers.filtered_top10 <- Lymphocytes.markers.filtered %>%
  group_by(cluster) %>%
  slice_max(avg_log2FC, n = 10)

DoHeatmap(object = Lymphocytes, features = Lymphocytes.markers.filtered_top10$gene, raster = F)

Idents(object = Lymphocytes) <- "seurat_clusters"
Lymphocytes<- RenameIdents(Lymphocytes, `0` = "CD8Ex T cells", `1` = "NK cells", `2` = "Stem-like T cells", `3` = "Cycling T cells", `4` = "Gzmk T cells", `5` = "Cycling T cells", `6` = "CD4 T cells")
Lymphocytes <- StashIdent(object = Lymphocytes, save.name = "cluster_name")

DoHeatmap(object = Lymphocytes, features = Lymphocytes.markers.filtered_top10$gene, raster = F)
DimPlot(Lymphocytes, reduction = "umap", group.by = "cluster_name", label = T)

```


```{r GSEA pathways setup}
# NOTE about GSEA: Negative enrichment in a cell type in scRNAseq data is meaningless because the FindMarkers function
collections <- msigdbr_collections()

Hallmark_df <- msigdbr(species = "Mus musculus", category = "H")
C7_df <- msigdbr(species = "Mus musculus", category = "C7")
C2_df <- msigdbr(species = "Mus musculus", category = "C2")
reactome_df <- msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:REACTOME")
C5_df <- msigdbr(species = "Mus musculus", category = "C5")
molecular_function_df <- msigdbr(species = "Mus musculus", category = "C5",subcategory = "GO:MF")
biological_process_df <- msigdbr(species = "Mus musculus", category = "C5",subcategory = "GO:BP")




Hallmark_pathways <-  split(x = Hallmark_df$gene_symbol, f = Hallmark_df$gs_name)
C7_pathways <- split(x = C7_df$gene_symbol, f = C7_df$gs_name)
C2_pathways <- split(x = C2_df$gene_symbol, f = C2_df$gs_name)
reactome_pathways <- split(x = reactome_df$gene_symbol, f = reactome_df$gs_name)
C5_pathways <- split(x = C5_df$gene_symbol, f = C5_df$gs_name)
molecular_function_pathways <- split(x = molecular_function_df$gene_symbol, f = molecular_function_df$gs_name)
biological_process_pathways <- split(x = biological_process_df$gene_symbol, f = biological_process_df$gs_name)

```

```{r}
Idents(object = Lymphocytes) <- "cluster_name"
CD8_Tcells <- subset(Lymphocytes, idents = c("CD8Ex T cells","Gzmk T cells"))
CD8_Tcells <- SetIdent(CD8_Tcells, value = CD8_Tcells@meta.data$model)
CD8_Tcell_markers <- FindMarkers(CD8_Tcells, ident.1 = "WT", ident.2 = "KO", assay = "RNA", test.use = "wilcox", logfc.threshold = 0.05)

CD8_Tcell_markers <- CD8_Tcell_markers %>% 
               rownames_to_column(var = "gene") %>% 
               dplyr::filter(!grepl("Trbv|Trav", gene))

CD8_Tcells_filt <- CD8_Tcell_markers %>%
                as.tibble() %>%
                dplyr::filter(-log10(p_val_adj) > 10) %>%
                dplyr::filter(abs(avg_log2FC) > 0.5)

ggplot(data=CD8_Tcell_markers, aes(x=avg_log2FC, y=-log10(p_val_adj))) + 
  xlim(-2,2.75) +
  geom_point(size = 1) + 
  ggtitle("Differentially Expressed Genes from CD8+ T cells") + 
  geom_point(data=CD8_Tcells_filt, colour="red", size=1) + 
  geom_text(aes(label = ifelse(-log10(p_val_adj) > 10 & abs(avg_log2FC) > 0.5, as.character(CD8_Tcells_filt$gene),'')), size=3, hjust=0, vjust=0, nudge_x = 0.05, check_overlap = T) +
  geom_vline(color = "darkgray",linetype = 4, xintercept = 0.5) +
  geom_vline(color = "darkgray",linetype = 4, xintercept = -0.5) +
  geom_hline(color = "darkgray",linetype = 4, yintercept = 10)


CD8_Tcells_top <- CD8_Tcell_markers %>% 
              as.tibble() %>%
              dplyr::select(gene, avg_log2FC) %>%
              dplyr::arrange(desc(abs(avg_log2FC))) %>%
              slice_head(n=30)
CD8_Tcells_top

#GSEA for CD8 T cells
CD8_Tcells_ranks <- CD8_Tcell_markers %>% 
                 dplyr::select(gene, avg_log2FC) %>% 
                 deframe()

gseaHallmarks_CD8_Tcells <- fgsea(pathways=Hallmark_pathways, stats=CD8_Tcells_ranks)
gseaC7_CD8_Tcells <- fgsea(pathways=C7_pathways, stats=CD8_Tcells_ranks)

gseaHallmarks_CD8_Tcells_filt <- gseaHallmarks_CD8_Tcells %>% slice_max(NES, n=10)
gseaHallmarks_CD8_Tcells_filt2 <- gseaHallmarks_CD8_Tcells %>% slice_min(NES, n=10)
gseaHallmarks_CD8_Tcells_filt <- rbind(gseaHallmarks_CD8_Tcells_filt,gseaHallmarks_CD8_Tcells_filt2) %>% mutate(pathway = gsub("HALLMARK_", "", pathway))


ggplot(gseaHallmarks_CD8_Tcells_filt, aes(reorder(pathway, NES), NES)) +
  theme(axis.text.y=element_text(size=6)) +
  geom_col(aes(fill=NES>0)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways for CD8 T Cells") + NoLegend()

gseaC7_CD8_Tcells_filt <- gseaC7_CD8_Tcells %>%
                as.tibble() %>%
                dplyr::filter((padj) < 0.05) %>%
                dplyr::filter((size) > 30) %>%
                dplyr::filter(NES > 2.3 | NES < -3) 

ggplot(gseaC7_CD8_Tcells_filt, aes(reorder(pathway, NES), NES)) +
  theme(axis.text.y=element_text(size=6)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  scale_x_discrete(label = function(x) stringr::str_trunc(x, 20)) +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="C7 Immune pathways for CD8 T Cells")

plotEnrichment(Hallmark_pathways[["HALLMARK_INTERFERON_GAMMA_RESPONSE"]],
               CD8_Tcells_ranks) + labs(title="Hallmark - Interferon gamma response", subtitle = "CD8+ T cells") + theme(plot.title = element_text(size = 11),plot.subtitle = element_text(size = 10)) 

```

